pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,300 @@
+}
+
+// Write ensures that the given payload is projected onto disk, updating the
+// contents of the AtomicWriter's target directory atomically if any writes
+// need to be performed.  A sentinel file named "..sentinel" in the target
+// directory is updated, allowing consumers of the directory to listen for
+// updates to the projected data by monitoring the sentinel file with inotify
+// or fanotify.
+//
+// The visible files in this volume are symlinks to files in the writer's
+// current state directory directory.  Actual files are stored in a hidden
+// timestamped directory which is symlinked to by the current state directory.
+// The timestamped directory and current state symlink are created in the
+// writer's target dir.  This scheme allows the files to be atomically updated by
+// changing the target of the current state symlink.
+//
+// When write is called:
+//
+// 0.  A check is made to determine whether the payload requires writing to disk.
+//     If the payload is already present on disk, the function returns.
+// 1.  A new timestamped dir is created by newTimestampDir
+// 2.  The payload is written to the new timestamped directory
+// 3.  Symlinks and directory for new files are created (if needed).
+//
+//     For example for files:
+//       <target-dir>/user_space/labels
+//       <target-dir>/k8s_space/annotations
+//       <target-dir>/podName
+//
+//     This structure is created:
+//       <target-dir>/podName               -> ..<identifier-token>/podName
+//       <target-dir>/user_space/labels     -> ../..<identifier-token>/user_space/labels
+//       <target-dir>/k8s_space/annotations -> ../..<identifier-token>/k8s_space/annotations
+//       <target-dir>/..<identifier-token>         -> ..<identifier-token>.12345678
+//
+//     where ..<identifier-token>.12345678 is a randomly generated directory that
+//     contains the real data.  If a file has to be dumped in subdirectory
+//     (for example <target-dir>/user_space/labels), the writer builds a
+//     relative symlink (<target-dir>/user_space/labels -> ../..<identifier-token>/user_space/labels)
+// 4.  The previous timestamped directory is detected reading the '..<identifier-token>' symlink
+// 5.  If no symlink for the current state directory exists, the link is created
+// 6.  If a symlink exists, a new temporary symlink is created: ..<identifier-token>_tmp
+// 7.  ..<identifier-token>_tmp is renamed to ..<identifier-token>
+// 8.  The previous timestamped directory is removed
+// 9.  The sentinel file is updated
+func (w *AtomicWriter) Write(payload map[string][]byte) error {
+	shouldWritePayload, err := w.shouldWritePayload(payload)
+	if err != nil {
+		glog.Errorf("Unable to determine whether payload should be written to disk: %v", err)
+		return err
+	}
+
+	if !shouldWritePayload {
+		glog.V(5).Infof("No update required for target directory %v", w.TargetDir)
+		return nil
+	}
+
+	tsDir, err := w.newTimestampDir()
+	if err != nil {
+		return err
```

Comment:
So, most of the other errors here seem to have an associated glog.Errorf message giving more detail as to the context.  Perhaps this should print out "Unable to created new payload data directory" or something of the sort?